add ability for gce to bulk verify attached disks

hantaowang opened the pull request:
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
The GCE Attacher is missing the bulk verify option. This PR adds the ability for GCE to bulk verify if disks are correctly attached.

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
```release-note
None
```

CC: @jingxu97 @saad-ali @msau42 @wojtek-t @cheftako @verult

Commit:
add ability for gce to bulk verify attached disks

msau42 commented:
/ok-to-test
/assign @davidz627

wojtek-t commented:
/assign

jingxu97 commented on staging/src/k8s.io/legacy-cloud-providers/gce/gce_disks.go:
```diff
@@ -620,6 +624,68 @@ func (g *Cloud) DisksAreAttached(diskNames []string, nodeName types.NodeName) (m
+					notFound = append(notFound, instanceName)
+				}
+			}
+			if err == cloudprovider.InstanceNotFound {
```
what is check? Different from line 643?

hantaowang replied:
Sorry, that was a mistake.

hantaowang replied:
Sorry, that was a mistake.

jingxu97 commented on staging/src/k8s.io/legacy-cloud-providers/gce/gce_disks.go:
```diff
@@ -620,6 +624,68 @@ func (g *Cloud) DisksAreAttached(diskNames []string, nodeName types.NodeName) (m
 	return attached, nil
 }
 
+// BulkDisksAreAttached is a batch function to check if all corresponding disks are attached to the
```
there are quite some loop functions  in this BulkDisksAreAttached function, I am wondering whether we can reduce some to improve the performance.

hantaowang replied:
While there are a lot of for loops, the run time should still be O(numNodes + numDisks). Some of the loops are a bit tricky - despite being a double for loop it still looks over only as many items as there are disks or nodes. 

Assuming that the runtime for getInstancesByName is constant. What's really going to slow it down is probably making API queries if getInstancesByName's heuristic is not good.

hantaowang commented:
/test pull-kubernetes-e2e-gce-100-performance

wojtek-t commented on pkg/volume/gcepd/attacher.go:
```diff
@@ -141,6 +141,47 @@ func (attacher *gcePersistentDiskAttacher) VolumesAreAttached(specs []*volume.Sp
+		deviceNames := []string{}
+		for _, spec := range volumeSpecs {
+			volumeSource, _, err := getVolumeSource(spec)
+			// If error is occurred, skip this volume and move to the next one
```
That comment doesn't provide any value - it describes what the code does.
Please remove.

wojtek-t commented on pkg/volume/gcepd/attacher.go:
```diff
@@ -141,6 +141,47 @@ func (attacher *gcePersistentDiskAttacher) VolumesAreAttached(specs []*volume.Sp
+
+	attachedDisksByNode, err := attacher.gceDisks.BulkDisksAreAttached(deviceNamesByNode)
+	if err != nil {
+		// Log error and continue with attach
```
Please remove this comment.

wojtek-t commented on pkg/volume/gcepd/attacher.go:
```diff
@@ -141,6 +141,47 @@ func (attacher *gcePersistentDiskAttacher) VolumesAreAttached(specs []*volume.Sp
+		klog.Errorf(
+			"Error bulk checking if PDs are already attached to nodes (%v). err=%v",
+			volumesByNode, err)
+		return volumesAttachedCheck, err
```
return nil, err

wojtek-t commented on staging/src/k8s.io/legacy-cloud-providers/gce/gce_disks.go:
```diff
@@ -620,6 +624,66 @@ func (g *Cloud) DisksAreAttached(diskNames []string, nodeName types.NodeName) (m
+	for nodeName, attachedDisks := range diskByNodes {
+		verifyDisksAttachedToNode := make(map[string]bool)
+
+		if disksActuallyAttached, ok := listedInstanceNamesToDisks[mapNodeNameToInstanceName(nodeName)]; ok {
```
can you opaque this into a separate function (this one is already quite long)
so sth like:
```
verifyDisksAttached[nodeName] = verifyDisksAttachedToNode(attachedDisks, disksActuallyAttached)
```

wojtek-t commented on staging/src/k8s.io/legacy-cloud-providers/gce/gce_instances.go:
```diff
@@ -480,12 +485,7 @@ func (g *Cloud) getInstancesByNames(names []string) ([]*gceInstance, error) {
-	var ret []*gceInstance
-	for _, instance := range found {
-		ret = append(ret, instance)
+		return ret, cloudprovider.InstanceNotFound
```
I'm wondering if it won't break some assuming somewhere... but hopefully not.

davidz627 replied:
oh I see now that this was modified. It feels weird to me to both return an error and a partial result back, then do some intense error checking to surface messages on the caller side.

What are the cases where an instance may not exist and why would we want to not just throw and error and let the caller of the `BulkVerify` function retry with the correct instances/nodes next time?

wojtek-t replied:
I think the intention here was to avoid situation when we list say 5k instances, 1 is not found and we drop everything.

But after thinking a bit more about it, I would much rather prefer refactoring this function (in the worst case introducing a new one) that would be returning some structured thing (rather than partial result and error).

So something like map[string]*gce.Instance (and treat nil as notFound) or map[string]struct{gce.instance, error}

Would be much more intuitive.

hantaowang replied:
I didn't want to make big changes to what this function returns because it is staged to a separate repo that can be used by many other people. What you suggest makes more sense, but I fear is not backwards compatible enough.

msau42 replied:
Would it be strange to create a new function?

hantaowang replied:
This new function would be very similar to the old one. I can see the new one being a wrapper on the old function, but I don't know if it makes sense to have one just to change the return value. With the current return, I do infer which instances the InstanceNotFound error refers to by cross listing it with the instances we expect.

davidz627 replied:
I would prefer there be a new function that's very similar. The old function could be a wrapper on the new function or both could be wrapper on a third function. Implementation is up to you but I don't think that we should be changing the guarantees of the original function (especially to something that is non-standard like partial results while throwing an error).

davidz627 replied:
I don't believe this function is exported so it can't be consumed elsewhere. So you could potentially just change the signature of this function and update all the callers

wojtek-t replied:
This is a private method and we are free to do whatever we want with it.
Please create a separate PR to change it to what we need (what I suggested above).

davidz627 commented on pkg/volume/gcepd/attacher.go:
```diff
@@ -141,6 +141,45 @@ func (attacher *gcePersistentDiskAttacher) VolumesAreAttached(specs []*volume.Sp
+		for _, spec := range volumeSpecs {
+			volumeSource, _, err := getVolumeSource(spec)
+			if err != nil {
+				klog.Errorf("Error getting volume (%q) source : %v", spec.Name(), err)
```
add some more contextual info here like "while bulkverifyvolumes-ing" or something. Also why is this error not bad enough that we would stop processing and instead ignore it

hantaowang replied:
I guess with a big cluster, you can for some reason have a malformed volume spec but it shouldn't prevent all the other ok ones from being checked. I can see how that can be an issue though, since if an error is not returned the malformed spec is never dealt with, but I think that is better than not checking any at all.

msau42 replied:
I'm unsure if this is a scenario that is possible for users to get into (and pass API validation).

hantaowang replied:
Do you know who originally wrote this for VolumesAreAttached? Maybe they can have a better explanation for why the log is included or if there is a scenario we're not thinking of.

hantaowang replied:
@jingxu97

gnufied replied:
The volume verification mechanism exists as a fail-safe to correct volumes that would have been incorrectly detached from nodes. So, it is *only* designed to cause attaches and never detaches. 

And since it verifies entire cluster it runs less frequently than other volume operations. So, if we come across a volume which we can't verify (like `getVolumeSource` failed!) then it is best to skip it rather than letting the whole operation fail.

jingxu97 replied:
@hantaowang I think my thought about it is same as yours. I think if our API validation does its job correctly, the error should never happen. Also, this function is used by other functions such as Attach(), if it returns error, attach will fail and verify function will not be called at all. Unless someone somehow modify the PV and remove its PD name after volume is attached, which would be really strange.

davidz627 commented on pkg/volume/gcepd/attacher.go:
```diff
@@ -141,6 +141,45 @@ func (attacher *gcePersistentDiskAttacher) VolumesAreAttached(specs []*volume.Sp
+
+	attachedDisksByNode, err := attacher.gceDisks.BulkDisksAreAttached(deviceNamesByNode)
+	if err != nil {
+		klog.Errorf(
```
does the err already get logged at a higher level, if so, we can just incoroporate this contextual info into the error returned and remove the logging here

hantaowang replied:
This is again something I was following from the current implementation of VolumesAreAttached. I am sure it gets handled by the operation executor / nested pending operation.

hantaowang replied:
I removed it, but what about the one on line 128?

davidz627 commented on staging/src/k8s.io/legacy-cloud-providers/gce/gce_disks.go:
```diff
@@ -620,6 +624,50 @@ func (g *Cloud) DisksAreAttached(diskNames []string, nodeName types.NodeName) (m
+	for _, instance := range listedInstances {
+		listedInstanceNamesToDisks[instance.Name] = instance.Disks
+	}
+	if err != nil {
```
this err check should probably occur before trying to use `listedInstances` right after getting the instance by name

davidz627 replied:
above for loop could potentially result in nil pointer derefence

davidz627 commented on staging/src/k8s.io/legacy-cloud-providers/gce/gce_disks.go:
```diff
@@ -620,6 +624,50 @@ func (g *Cloud) DisksAreAttached(diskNames []string, nodeName types.NodeName) (m
+	if err != nil {
+		if err == cloudprovider.InstanceNotFound {
+			notFound := []string{}
+			for _, instanceName := range instanceNames {
```
not sure what this inner error loop is doing, `getInstanceByNames` always returns nil for the list of instances if it returns an error

hantaowang replied:
I changed that function so that if the error is cloudprovider.InstanceNotFound, it doesn't return nil but rather the list of instances it could find.

hantaowang commented:
/assign jingxu97

davidz627 commented on staging/src/k8s.io/legacy-cloud-providers/gce/gce_disks.go:
```diff
@@ -990,3 +1038,22 @@ func isGCEError(err error, reason string) bool {
+// array of disks actually attached to it. It returns a verifying if the disks are actually attached.
+func verifyDisksAttachedToNode(disksToVerify []string, disksActuallyAttached []*compute.AttachedDisk) map[string]bool {
+	verifiedDisks := make(map[string]bool)
+	disksActuallyAttachedSet := make(map[string]struct{})
```
use `sets.String`. you can directly change the list into a Set and use set `contains` to check for existence

hantaowang replied:
I don't think sets.String can easily be used, because the array passed in is not an array of strings but an array of AttachedDisks item. I would have to apply a map that gets the name of each AttachedDisk but the cost of that is another loop which will only make the code more confusing.

wojtek-t replied:
use the same loop as you have before (+1 for using sets.String)

wojtek-t commented on staging/src/k8s.io/legacy-cloud-providers/gce/gce_disks.go:
```diff
@@ -620,6 +624,49 @@ func (g *Cloud) DisksAreAttached(diskNames []string, nodeName types.NodeName) (m
+	// Then for each instance listed, add the disks attached to that instance to a map
+	listedInstances, err := g.getInstancesByNames(instanceNames)
+	listedInstanceNamesToDisks := make(map[string][]*compute.AttachedDisk)
+	if err != nil && err != cloudprovider.InstanceNotFound {
```
Handle error first (before allocating listedInstanceNamesToDisks)

hantaowang replied:
I moved it after the generic error handling but before the cloudprovider.InstanceNotFound handling because for InstanceNotFound to log the not found instances I need listedInstanceNamesToDisks.

wojtek-t commented on staging/src/k8s.io/legacy-cloud-providers/gce/gce_disks.go:
```diff
@@ -620,6 +624,49 @@ func (g *Cloud) DisksAreAttached(diskNames []string, nodeName types.NodeName) (m
+	for nodeName, disksToVerify := range diskByNodes {
+		if disksActuallyAttached, ok := listedInstanceNamesToDisks[mapNodeNameToInstanceName(nodeName)]; ok {
+			verifyDisksAttached[nodeName] = verifyDisksAttachedToNode(disksToVerify, disksActuallyAttached)
+		} else {
```
You don't need else - verifyDisksAttachedToNode should accept nil parameters and then you can just do:

verifyDisksAttached[nodeName] = verfiDisksAttachedToNode(disksToVerify, listedInstancesNamesToDisks[...])

hantaowang commented on staging/src/k8s.io/legacy-cloud-providers/gce/gce_disks.go:
```diff
@@ -990,3 +1033,23 @@ func isGCEError(err error, reason string) bool {
+func verifyDisksAttachedToNode(disksToVerify []string, disksActuallyAttached []*compute.AttachedDisk) map[string]bool {
+	verifiedDisks := make(map[string]bool)
+	diskNamesActuallyAttached := []string{}
+	if disksActuallyAttached != nil {
```
what part of this is unnecessary?

hantaowang commented:
Thanks everyone for the comments. I've fixed or responded to most of them and now squashed all the intermediate commits. The only outstanding thing is how we deal with cloudprovider.InstanceNotFound, which I will follow the suggestions of @wojtek-t and open a separate PR for. 

If there is anything else, please add more reviews otherwise I think it can ready for final review + approval.

hantaowang commented:
/remove-sig api-machinery apps auth autoscaling cli cluster-lifecycle node release scheduling testing
/remove-area apiserver dependency e2e-test-framework kubeadm kubectl kubelet test

hantaowang commented:
Sorry all for the confusion - my rebase and force push made everything go haywire. I'll repeat my earlier comment below in case it got lost in the hidden items. 

---
Thanks everyone for the comments. I've fixed or responded to most of them and now squashed all the intermediate commits. The only outstanding thing is how we deal with cloudprovider.InstanceNotFound, which I will follow the suggestions of @wojtek-t and open a separate PR for.

If there is anything else, please add more reviews otherwise I think it can ready for final review + approval.

hantaowang commented:
/hold

wojtek-t commented on staging/src/k8s.io/legacy-cloud-providers/gce/gce_instances.go:
```diff
@@ -424,54 +425,68 @@ func (g *Cloud) AddAliasToInstance(nodeName types.NodeName, alias *net.IPNet) er
-func (g *Cloud) getInstancesByNames(names []string) ([]*gceInstance, error) {
+// instance is not found. Always returns the instances that could be found.
+// Maximum names per request is set to the default page size of 500.
+func (g *Cloud) getInstancesByNames(allNames []string) ([]*gceInstance, error) {
```
You still didn't change the interface and I definitely agree that partial-result+error is a bad thing to use.

Can you please open a separate PR and do whatever changes are needed for this function for the purpose of this PR there?

hantaowang replied:
I just opened https://github.com/kubernetes/kubernetes/pull/80180. I'll rebase this when that is merged.

hantaowang commented:
/uncc gnufied mikedanese

wojtek-t commented on staging/src/k8s.io/legacy-cloud-providers/gce/gce_disks.go:
```diff
@@ -620,6 +624,36 @@ func (g *Cloud) DisksAreAttached(diskNames []string, nodeName types.NodeName) (m
+
+	// For each node and its desired attached disks that needs to be verified
+	for nodeName, disksToVerify := range diskByNodes {
+		disksActuallyAttached := listedInstanceNamesToDisks[canonicalizeInstanceName(mapNodeNameToInstanceName(nodeName))]
```
nit:
```
instanceName := canonicalizeInstanceName(mapNodeNameToInstanceName(nodeName))
disksActuallyAttached := listedInstanceNamesToDisks[instanceName]
verifyDisksAttached[nodeName] = verifyDisksAttachedToNode(disksToVerify, disksActuallyAttached)
```

to avoid such long lines

hantaowang replied:
ack

wojtek-t commented on staging/src/k8s.io/legacy-cloud-providers/gce/gce_disks.go:
```diff
@@ -990,3 +1024,21 @@ func isGCEError(err error, reason string) bool {
+	verifiedDisks := make(map[string]bool)
+	diskNamesActuallyAttached := []string{}
+	for _, disk := range disksActuallyAttached {
+		diskNamesActuallyAttached = append(diskNamesActuallyAttached, disk.DeviceName)
```
I don't get it - why we create a slice first. Should be:
```
diskNameActuallyAttached := sets.NewString()
for _, disk := range disksActuallyAttached {
  disksNamesActuallyAttached.Insrt(disk.DeviceName)
}
```

hantaowang replied:
ack

wojtek-t reviewed (commented):
Two minor comments - and I think we should merge it asap.

hantaowang commented:
/hold cancel
/test pull-kubernetes-integration

wojtek-t commented:
/lgtm
/approve

tedyu commented on staging/src/k8s.io/legacy-cloud-providers/gce/gce_disks.go:
```diff
@@ -620,6 +624,37 @@ func (g *Cloud) DisksAreAttached(diskNames []string, nodeName types.NodeName) (m
+
+	// List all instances with the given instance names
+	// Then for each instance listed, add the disks attached to that instance to a map
+	listedInstances, err := g.getFoundInstanceByNames(instanceNames)
```
It seems we only use disk names of the return value from Cloud#getFoundInstanceByNames .
Instead of mapping to []*compute.AttachedDisk, listedInstanceNamesToDisks can map to Set of disk names which verifyDisksAttachedToNode can directly utilize ?

@jingxu97 @hantaowang what do you think ?
